// easynodeapi/NodeAPI.js
var fs = require("fs");
var qs = require("querystring");
console.log("Start API");
module.exports = function (req,res) {
    ServerFunc = [];
    return {
    res:res,
    req:req,
    ViewLang:function(static,page) {
        var data = fs.readdir(static,function(err,files){
            if(err){
                console.log("error ",err);
            }
            else
            {
                var alang = req.headers["accept-language"]
                for(var i=0;i<files.length;i++)
                {
                    if(alang.indexOf(files[i]) > -1)
                    {
                        console.log(static+"/"+files[i]+"/" + page);
                        fs.readFile(static+"/"+files[i]+"/" + page, function (error, data) {
                            if (error) {
                                res.write("Error");
                                res.end();
                            }
                            else {
                                res.writeHead(200, { 'Content-Type': 'text/html' });
                                res.write(data);
                                res.end();
                                return;
                            }
                        })
                    }
                }
            }
        })
    },
    GetFile : function(file,callback){
        fs.readFile(file, function (error, data) {
            if (error) {
                return "Error " + error;
            }
            else {
                callback(data);
            }
        })
    },
    View:function(file)
    {
        fs.readFile(file, function (error, data) {
            if (error) {
                res.write("Error ",error);
                res.end();
            }
            else {
                res.writeHead(200, { 'Content-Type': 'text/html' });
                res.write(data);
                res.end();
            }
        })
    },
    File:function(file)
    {
        var roz = file.substring(file.indexOf(".",1)+1,file.length);
        fs.readFile(file, function (error, data) {
            if (error) {
                res.write("Error ",error);
                console.log(error.message);
                res.end();
            }
            else {
                switch(roz){
                    case "html":
                        res.writeHead(200, { 'Content-Type': 'text/html' });
                        break;
                    case "css":
                        res.writeHead(200, { 'Content-Type': 'text/css' });
                        break;
                    case "jpeg":
                        res.writeHead(200, { 'Content-Type': 'image/jpeg' });
                        break;
                    case "png":
                        res.writeHead(200, { 'Content-Type': 'image/png' });
                        break;
                    case "xml":
                        res.writeHead(200, { 'Content-Type': 'text/xml' });
                        break;
                    default:
                        res.writeHead(200, { 'Content-Type': 'text/plain' });
                        break;
                }
                res.write(data);
                res.end();
            }
        })
    },
    Json:function(data)
    {
        res.writeHead(200, { 'Content-Type': "application/json" })
        res.write(JSON.stringify(data));
        res.end();
    },
    SendPOSTData:function(func)
    {
        var str = "";
        var obj = {};
        var dat = new Date();
        req.on("data", function (data) {
            str += data;
        })
        req.on("end", function (data) {
            obj = qs.parse(str);
            if (func)
            {
                func(obj);
            }
            res.end(JSON.stringify(obj));
        })
    },
    GetPOSTData:function(func){
        var str = "";
        var obj = {};
        var dat = new Date();
        req.on("data", function (data) {
            str += data;
        })
        req.on("end", function (data) {
            obj = qs.parse(str);
            if (func)
            {
                func(obj)
            }
        })
    },
    SavePOSTData: function (name,func) {
        var str = "";
        var obj = {};
        var dat = new Date();
        req.on("data", function (data) {
            str += data;
        })
        req.on("end", function (data) {
            obj = qs.parse(str);
            if (func)
            {
                func(obj)
            }
            this.Data[name] = obj;
        })
    },
    OnGETUrl: function (url, func) {
        var reqUrl = UrlFilter(url);
        ServerFunc.push({url:reqUrl,method:"GET",func:func});
    },
    OnPOSTUrl: function ( url, func) {
        var reqUrl = UrlFilter(url);
        ServerFunc.push({url:reqUrl,method:"POST",func:func});
    },
    OnUrl:function(method,url,func){
        var reqUrl = UrlFilter(url);
        ServerFunc.push({url:reqUrl,method:method,func:func});
    },
    Default:function(func){
        this.ServerDefault = func;
    },
    Tab:{
        RemFromArray:function (index,table)
        {
            if(index = table.length)
            {
                table.pop()
                return table;
            }
            else
            {
                for(var i=index;i<table.length-1;i++)
                {
                    table[index] = table[index + 1];
                }
                table.pop();
                return table;
            }
        }
    },
    executedReq:false,
    ServerDefault:null,
    Server:function(){
        if(!this.ServerDefault)
        {
            console.log("You must set the default response");
        }
        for(var i=0;i<ServerFunc.length;i++)
        {
            if (this.req.url == ServerFunc[i].url && this.req.method == ServerFunc[i].method)
            {
                ServerFunc[i].func();
                this.executedReq = true;
                return;
            }
        }
        if(!this.executedReq)
        {
            this.ServerDefault();
        }
    }
    }
}
function AddToObject(obj,name,value)
{
    eval("obj." + name + " = " + JSON.stringify(value));
    return obj;
}
function UrlFilter(url)
{
    var ret = "";
    for(var i=0;i<url.length;i++)
    {
        if(url[i] == "?")
        {
            return ret;
        }
        else
        {
            ret += url[i];
        }
    }
    return url;
}
